consensus/src/experimental/tests/commit_phase_tests.rs

```diff
@@ -0,0 +1,383 @@
+    let (signers, validators) = random_validator_verifier(num_nodes, None, false);
+    let validator_set = (&validators).into();
+    let signer = &signers[0];
+
+    let waypoint =
+        Waypoint::new_epoch_boundary(&LedgerInfo::mock_genesis(Some(validator_set))).unwrap();
+
+    let safety_storage = PersistentSafetyStorage::initialize(
+        Storage::from(diem_secure_storage::InMemoryStorage::new()),
+        signer.author(),
+        signer.private_key().clone(),
+        Ed25519PrivateKey::generate_for_testing(),
+        waypoint,
+        true,
+    );
+    let safety_rules_manager = SafetyRulesManager::new_local(safety_storage, false, false, true);
+
+    let (_initial_data, storage) = MockStorage::start_for_testing((&validators).into());
+    let epoch_state = EpochState {
+        epoch: 1,
+        verifier: storage.get_validator_set().into(),
+    };
+    let validators = epoch_state.verifier.clone();
+    let (network_reqs_tx, _network_reqs_rx) = diem_channel::new(QueueStyle::FIFO, 8, None);
+    let (connection_reqs_tx, _) = diem_channel::new(QueueStyle::FIFO, 8, None);
+
+    let network_sender = ConsensusNetworkSender::new(
+        PeerManagerRequestSender::new(network_reqs_tx),
+        ConnectionRequestSender::new(connection_reqs_tx),
+    );
+    let author = signer.author();
+
+    let (self_sender, self_receiver) = channel::new_test(1000);
+    let network = NetworkSender::new(author, network_sender, self_sender, validators);
+
+    let (commit_result_sender, commit_result_receiver) =
+        channel::new_test::<(Vec<Block>, LedgerInfoWithSignatures)>(channel_size);
+    let state_computer = Arc::new(OrderingStateComputer::new(commit_result_sender));
+
+    let mut safety_rules = MetricsSafetyRules::new(safety_rules_manager.client(), storage);
+    safety_rules.perform_initialize().unwrap();
+
+    let safety_rules_container = Arc::new(Mutex::new(safety_rules));
+
+    // setting up channels
+    let (sender_comm, receiver_comm) =
+        channel::new_test::<(Vec<ExecutedBlock>, LedgerInfoWithSignatures)>(channel_size);
+
+    let (sender_c_msg, receiver_c_msg) = channel::new_test::<ConsensusMsg>(channel_size);
+
+    let commit_phase = CommitPhase::new(
+        receiver_comm,
+        state_computer.clone(),
+        receiver_c_msg,
+        epoch_state.verifier.clone(),
+        safety_rules_container.clone(),
+        author,
+        back_pressure,
+        network,
+    );
+
+    runtime.spawn(commit_phase.start());
+    (
+        sender_comm,            // channel to pass executed blocks into the commit phase
+        sender_c_msg,           // channel to pass commit messages into the commit phase
+        commit_result_receiver, // channel to receive commit result from the commit phase
+        self_receiver,          // channel to receive message from the commit phase itself
+        safety_rules_container,
+        signers,
+        state_computer,
+        epoch_state.verifier,
+        runtime,
```

Comment:
it's probably better to pass the runtime in instead out in case we need to reuse a runtime to create multiple ones

Reply:
resolved